src/GoogleDriveAdapter.php

```diff
@@ -1086,117 +1086,164 @@ protected function createDirectory($name, $parentId)
+            $chunkSizeBytes = max(262144, min($chunkSizeBytes, floor($availableMemory / 262144) * 262144 / 4));
         }
 
-        $isResource = false;
-        if (is_resource($contents)) {
-            $fstat = @fstat($contents);
-            if (! empty($fstat['size'])) {
-                $isResource = true;
-            }
-            if (! $isResource) {
-                $contents = stream_get_contents($contents);
-            }
+        return (int)$chunkSizeBytes;
+    }
+
+    /**
+     * Normalise a Drive File that has been created
+     *
+     * @param Google_Service_Drive_DriveFile $uploadedFile
+     * @param string $localPath
+     * @param string $visibility
+     * @return array|bool
+     */
+    protected function normaliseUploadedFile($uploadedFile, $localPath, $visibility)
+    {
+        list ($parentId, $fileName) = $this->splitPath($localPath);
+
+        if (!($uploadedFile instanceof Google_Service_Drive_DriveFile)) {
+            return false;
         }
 
-        if ($isResource) {
-            // set chunk size (max: 100MB)
-            $chunkSizeBytes = 100 * 1024 * 1024;
-            $memory = $this->getIniBytes('memory_limit');
-            if ($memory > 0) {
-                $chunkSizeBytes = max(262144 , min([
-                    $chunkSizeBytes,
-                    (intval($memory / 4 / 256) * 256)
-                ]));
-            }
-            if ($fstat['size'] < $chunkSizeBytes) {
-                $isResource = false;
-                $contents = stream_get_contents($contents);
-            }
+        $this->cacheFileObjects[$uploadedFile->getId()] = $uploadedFile;
+        if (! $this->getFileObject($localPath)) {
+            $this->cacheFileObjectsByName[$parentId . '/' . $fileName] = $uploadedFile;
+        }
+        $result = $this->normaliseObject($uploadedFile, Util::dirname($localPath));
+
+        if ($visibility && $this->setVisibility($localPath, $visibility)) {
+            $result['visibility'] = $visibility;
         }
 
+        return $result;
+    }
+
+    /**
+     * Upload a PHP resource stream to Google Drive
+     *
+     * @param resource $resource
+     * @param string $localPath
+     * @param string $mime
+     * @return bool|Google_Service_Drive_DriveFile
+     */
+    protected function uploadResourceToGoogleDrive($resource, $localPath, $mime)
+    {
+        $fileSize = $this->getFileSizeBytes($resource);
+        $chunkSizeBytes = $this->detectChunkSizeBytes();
+
+        if ($fileSize <= $chunkSizeBytes) {
+            // If the resource fits in a single chunk, we'll just upload it in a single request
+            return $this->uploadStringToGoogleDrive(stream_get_contents($resource), $localPath, $mime);
+        }
+
+        // Call the API with the media upload, defer so it doesn't immediately return.
+        $request = $this->ensureDriveFileExists('', $localPath, $mime);
+        $client = $this->service->getClient();
+        $client->setDefer(true);
```

Comment:
@nao-pon
Thanks again for the great catch! Also, my apologies; I clearly didn't test this enough manually.

I've made a change to fix the issue you describe and tested that the code works with:
- A resource that doesn't fit into the available memory
- A resource that does fit into the available memory
- A string

Reply:
I tried your new version, and i got error "Catchable fatal error: Argument 2 passed to Google_Http_MediaFileUpload::__construct() must implement interface Psr\Http\Message\RequestInterface, instance of Google_Service_Drive_DriveFile given, called in vendor/nao-pon/flysystem-google-drive/src/GoogleDriveAdapter.php on line 1339 and defined in vendor/google/apiclient/src/Google/Http/MediaFileUpload.php on line 78".

It may be due to calling "getFileObject ()" function in "ensureDriveFileExists ()" function.

Reply:
@nao-pon Thanks for all the quick feedback! Sadly, I'm having trouble replicating this particular issue. The error you show is the one I was getting before my third commit.

After my third commit, the result returned from "ensureDriveFileExists" should be the same as the results of lines 1140/1144 of current nao-pon master branch.

Any thoughts on what I need to change locally to replicate the issue?

Reply:
Fixed with 789922b6f1a68e25d28481a592801343bac4cd25